textx/model.py

```diff
@@ -289,6 +301,85 @@ def get_model_from_str(self, model_str, file_name=None, debug=None,
+        def _hasattr(obj, name):
+            try:
+                return name in obj._tx_obj_attrs[id(obj)]
+            except KeyError:
+                if obj._tx_real_hasattr:
+                    return obj._tx_real_hasattr(name)
+                else:
+                    return super(type(obj), obj).__hasattr__(name)
+            except AttributeError:
+                return False
+
+        def _getattr(obj, name):
+            try:
+                return obj._tx_obj_attrs[id(obj)][name]
+            except KeyError:
+                if obj._tx_real_getattr:
+                    return obj._tx_real_getattr(name)
+                else:
+                    return super(type(obj), obj).__getattr__(name)
+
+        def _setattr(obj, name, value):
+            try:
+                obj._tx_obj_attrs[id(obj)][name] = value
+            except KeyError:
+                if obj._tx_real_setattr:
+                    return obj._tx_real_setattr(name, value)
+                else:
+                    return super(type(obj), obj).__setattr__(name, value)
+
+        def _delattr(obj, name):
+            try:
+                obj._tx_obj_attrs[id(obj)].pop(name)
+            except KeyError:
+                if obj._tx_real_delattr:
+                    return obj._tx_real_delattr(name)
+                else:
+                    return super(type(obj), obj).__delattr__(name)
+
+        def _replace_user_attr_methods(self):
+            """
+            Replace has/get/set/del(attr) methods on user classes
+            to support postponing of user obj initialization.
+            """
+            for user_class in self.metamodel.user_classes.values():
+                if not hasattr(user_class, '_tx_instrumented'):
+                    for a_name in ('has', 'get', 'set', 'del'):
+                        real_name = '__{}attr__'.format(a_name)
+                        cached_name = '_tx_real_{}attr'.format(a_name)
+                        setattr(user_class, cached_name,
+                                getattr(user_class, real_name, None))
+                        setattr(user_class, real_name,
+                                getattr(self.__class__,
+                                        '_{}attr'.format(a_name)))
+                    user_class._tx_instrumented = 1
+                else:
+                    user_class._tx_instrumented += 1
+
+        def _restore_user_attr_methods(self):
+            """
+            Restore original has/get/set/del(attr) methods on user
+            classes.
+            """
+            for user_class in self.metamodel.user_classes.values():
+                if hasattr(user_class, '_tx_instrumented'):
+                    user_class._tx_instrumented -= 1
+                    if user_class._tx_instrumented == 0:
+                        delattr(user_class, '_tx_instrumented')
+                        for a_name in ('has', 'get', 'set', 'del'):
```

Comment:
There is no `__hasattr__` hook in python, `hasattr` simply calls `getattr` and returns `False` if an `AttributeError` is thrown, and `True` otherwise. So there is no need to monkey patch it.

BTW, this also means that when running `if hasattr(obj, name) and getattr(obj, name) is not None` `obj.__getattr__` gets called twice, which could be avoided using a `try` `except` block, thought this might make the code read a little more clunky.

Reply:
Ah, you are right, good catch. Fixed.